File: apps/server/src/adapter/driven/email/resend.ts
import { Resend } from "resend";
import type { EmailService } from "../../../domain/port/index.js";

export interface ResendEmailConfig {
  apiKey: string;
  fromAddress: string;
}

export class ResendEmailService implements EmailService {
  private readonly client: Resend;
  private readonly from: string;

  constructor(config: ResendEmailConfig) {
    this.client = new Resend(config.apiKey);
    this.from = config.fromAddress;
  }

  async sendEmail(to: string, subject: string, html: string): Promise<void> {
    await this.client.emails.send({
      from: this.from,
      to,
      subject,
      html,
    });
  }

  async sendTemplateEmail(
    to: string,
    templateId: string,
    variables: Record<string, string>,
  ): Promise<void> {
    // Resend does not have a native template-by-ID endpoint like some
    // providers. We fetch the template content and interpolate variables
    // manually. For production you could use React Email components or
    // a stored-template mapping. Here we send with the subject derived
    // from the templateId and variables rendered into a simple HTML body.
    let html = `<p>Template: ${templateId}</p>`;
    for (const [key, value] of Object.entries(variables)) {
      html += `<p><strong>${key}:</strong> ${value}</p>`;
    }

    await this.client.emails.send({
      from: this.from,
      to,
      subject: templateId,
      html,
    });
  }
}

File: apps/api/internal/adapter/driven/email/resend.go
package email

import (
	"context"
	"fmt"

	"github.com/neurodynecorp/api/internal/domain/port"
	"github.com/neurodynecorp/api/pkg/logger"
	"github.com/resend/resend-go/v2"
)

type ResendService struct {
	client      *resend.Client
	fromAddress string
	log         *logger.Logger
}

func NewResendService(apiKey, fromAddress string, log *logger.Logger) *ResendService {
	return &ResendService{
		client:      resend.NewClient(apiKey),
		fromAddress: fromAddress,
		log:         log.WithComponent("email"),
	}
}

func (s *ResendService) SendEmail(ctx context.Context, to, subject, htmlBody string) error {
	params := &resend.SendEmailRequest{
		From:    s.fromAddress,
		To:      []string{to},
		Subject: subject,
		Html:    htmlBody,
	}

	sent, err := s.client.Emails.Send(params)
	if err != nil {
		s.log.Error().Err(err).Str("to", to).Str("subject", subject).Msg("failed to send email")
		return err
	}

	s.log.Info().Str("id", sent.Id).Str("to", to).Msg("email sent")
	return nil
}

func (s *ResendService) SendTemplateEmail(ctx context.Context, to, subject, templateID string, data map[string]interface{}) error {
	// For now, render a simple template. In production, use a proper template engine.
	html := fmt.Sprintf("<html><body><h1>%s</h1></body></html>", subject)
	return s.SendEmail(ctx, to, subject, html)
}

var _ port.EmailService = (*ResendService)(nil)
